velox/expression/tests/VectorReaderTest.cpp

```diff
@@ -0,0 +1,407 @@
+      [](vector_size_t idx) { return idx % 6 == 0; });
+  BufferPtr indices =
+      AlignedBuffer::allocate<vector_size_t>(size, execCtx_.pool());
+  auto rawIndices = indices->asMutable<vector_size_t>();
+  for (auto i = 0; i < size; ++i) {
+    // 13 and 10 are coprime so this shuffles the indices.
+    rawIndices[i] = (i * 13) % 10;
+  }
+  auto arrayVector = wrapInDictionary(indices, size, baseVector);
+
+  DecodedVector decoded;
+  exec::VectorReader<Array<int32_t>> reader(
+      decode(decoded, *arrayVector.get()));
+
+  // Empty Array.
+  ASSERT_FALSE(reader.containsNull(0));
+  // Non-empty arrays without NULLs.
+  ASSERT_FALSE(reader.containsNull(1));
+  ASSERT_FALSE(reader.containsNull(6));
+  // Array is NULL.
+  ASSERT_TRUE(reader.containsNull(4));
+  // Array contains NULL value.
+  ASSERT_TRUE(reader.containsNull(8));
+
+  // Next contains NULL.
+  ASSERT_FALSE(reader.containsNull(6, 7));
+  // Previous array is NULL.
+  ASSERT_FALSE(reader.containsNull(5, 6));
+  // Surrounded by NULL and array containing NULL.
+  ASSERT_FALSE(reader.containsNull(5, 7));
+  // NULL in middle of range.
+  ASSERT_TRUE(reader.containsNull(2, 6));
+  // First array is NULL.
+  ASSERT_TRUE(reader.containsNull(4, 7));
+  // Last array contains NULL.
+  ASSERT_TRUE(reader.containsNull(5, 8));
+}
+
+TEST_F(VectorReaderTest, rowContainsNull) {
+  // Vector is:
+  // [
+  //   {field1: NULL, field2: 0},
+  //   {field1: 1, field2: NULL},
+  //   {field1: 2, field2: 2},
+  //   NULL,
+  //   {field1: 4, field2: 4},
+  //   {field1: NULL, field2: 5},
+  //   {field1: 6, field2: NULL},
+  //   {field1: 7, field2: 7},
+  //   NULL,
+  //   {field1: 9, field2: 9},
+  // ]
+  auto field1Vector = makeFlatVector<int32_t>(
+      10,
+      [](vector_size_t row) { return row; },
+      [](vector_size_t row) { return row % 5 == 0; });
+  auto field2Vector = makeFlatVector<float>(
+      10,
+      [](vector_size_t row) { return row; },
+      [](vector_size_t row) { return row % 5 == 1; });
+  auto vector = makeRowVector(
+      {field1Vector, field2Vector},
+      [](vector_size_t idx) { return idx % 5 == 3; });
+  DecodedVector decoded;
+  exec::VectorReader<Row<int32_t, float>> reader(
+      decode(decoded, *vector.get()));
+
+  // Row without NULLs.
+  ASSERT_FALSE(reader.containsNull(4));
+  ASSERT_FALSE(reader.containsNull(7));
+  // Row is NULL.
+  ASSERT_TRUE(reader.containsNull(3));
+  // Row contains NULL field.
+  ASSERT_TRUE(reader.containsNull(1));
+
+  // Next row is NULL.
+  ASSERT_FALSE(reader.containsNull(2, 3));
+  // Previous row contains NULL.
+  ASSERT_FALSE(reader.containsNull(7, 8));
+  // Surrounded by NULL and row containing NULL.
+  ASSERT_FALSE(reader.containsNull(4, 5));
+  // NULL in middle of range.
+  ASSERT_TRUE(reader.containsNull(7, 10));
+  // First row contains NULL.
+  ASSERT_TRUE(reader.containsNull(1, 3));
+  // Last row contains NULL.
+  ASSERT_TRUE(reader.containsNull(4, 6));
+}
+
```

Comment:
maybe add a test for dictionary or constant encoded Row to catch the current issue.